src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -127,6 +164,55 @@ public void AddEntriesFromPreviousTable(NodeStateTable<T> previousTable, EntrySt
                 _states.Add(previousEntries);
             }
 
+            public void ModifyEntriesFromPreviousTable(NodeStateTable<T> previousTable, ImmutableArray<T> outputs)
+            {
+                // Semantics:
+                // For every slot in the previous table, we compare the new value.
+                // - Cached when the same
+                // - Modified when different
+                // - Removed when i > outputs.length
+                // - Added when i < previousTable.length
+
+                var previousEntries = previousTable._states[_states.Count];
```

Comment:
Rather than throwing here consider adding a `Debug.Assert` that validates `_states.Count` is a valid index and add a note / comment about what build invariant is being violated when than assert is tripped.